tests/test_spawn.py

```diff
@@ -0,0 +1,94 @@
+import os
+import requests
+import pytest
+import time
+
+# Makes heavy use of JupyterHub's API:
+# http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyterhub/jupyterhub/master/docs/rest-api.yml
+
+def test_api(request_data):
+    print("asking for the hub's version")
+    r = requests.get(request_data['hub_url'])
+    r.raise_for_status()
+    print(r.json())
+    return r.json()
+
+def test_api_info(request_data):
+    print("asking for the hub information")
+    r = requests.get(request_data['hub_url'] + '/info', headers=request_data['headers'])
+    r.raise_for_status()
+    print(r.json())
+    return r.json()
+
+def test_api_create_user(request_data):
```

Comment:
maybe move all the `test_api_...` functions into a new `test_api.py` file and rename them `test_create_user` instead of `test_api_create_user`?

Reply:
Sure! It is the first time i setup to use pytest myself and have lots to learn.

There is a lot of state relating to these tests. Should I ensure all tests will return the cluster to the original state? If i do so I'll need to make very few and very large tests though.

If i don't ensure state is restored in between tests, I need to keep track of the execution order. I learned that the test functions will be called in the order I've declared them within the file, but what about having multiple files, in what order will they execute? Will google this soon

Reply:
Ok. I hadn't considered that the functions depend on each other. Extrapolating from myself I would say that test functions that depend on each other isn't something we should do as no one would expect that to be the case. (At least setup in a vanilla way like this)

There are helpers in pytest to setup/undo state for a test function and reuse that tooling across many test functions. Maybe something to tackle in a new PR though. That way others with a lot of pytest experience can also get involved. Let's see what others think.

Reply:
We can improve this with some useful pytest fixtures, likely some module- and session-scope fixtures. Relying on test order is something one generally tries to avoid, since it makes isolating tests and failures difficult. That can be improved in another PR, though.

Reply:
I would also love for the test to be independent and be runable without any state to consider. This is not testing of Python code itself that can be isolated so easily though, it is testing of a kubernetes cluster behavior when running a few webrequests to a JupyterHub server running in the cluster.

Yes it would be **great to have someone more experienced** look at it later. I have never setup pytest myself before, my **experienced** friend @heyhugo got me started yesterday! :D